include/FastBVH/BuildStrategy1.h

```diff
@@ -0,0 +1,182 @@
+#include <FastBVH/BuildStrategy.h>
+
+namespace FastBVH {
+
+//! \brief Contains details on the implementation
+//! of the variant-1 BVH build strategy.
+namespace Strategy1 {
+
+//! \brief Contains the context used while building
+//! a specific node in the BVH.
+struct BuildEntry final {
+  //! If non-zero then this is the index of the parent. (used in offsets)
+  uint32_t parent;
+  //! The starting index of the range of primitives in this node.
+  uint32_t start;
+  //! The ending index of the range of primitives in this node.
+  uint32_t end;
+};
+
+//! \brief Used while constructing the BVH
+//! to queue nodes to be built.
+class BuildStack final {
+  //! The maximum number of entries in the stack.
+  static constexpr std::size_t max_size = 128;
+  //! The entry array of the stack.
+  BuildEntry entries[max_size];
+  //! The index of the current entry, plus one.
+  //! A value of zero indicates an empty stack.
+  std::size_t stack_ptr = 0;
+public:
+  //! Pops the top entry from the stack.
+  //! This function does not check if the
+  //! stack is empty.
+  auto pop() noexcept {
+    return entries[--stack_ptr];
+  }
+  //! Pushes an entry to the stack.
+  //! \param entry The entry to be pushed.
+  void push(const BuildEntry& entry) noexcept {
+    entries[stack_ptr] = entry;
+    stack_ptr++;
+  }
+  //! Indicates the size of the stack.
+  //! \return The number of entries in the stack.
+  //! A value of zero indicates the stack is empty.
+  auto size() const noexcept {
+    return stack_ptr;
+  }
+  //! Accesses an entry from the stack, by index.
+  //! \param index The index of the entry to get.
+  //! This is not a top-down index, so an index of
+  //! zero would reference the first item added to the stack.
+  //! \return A reference to the specified entry.
+  const auto& operator [] (std::size_t index) const noexcept {
+    return entries[index];
+  }
+};
+
+} // namespace Strategy1
+
+template <typename Float>
+template <typename Primitive, typename BoxConverter>
+BVH<Float, Primitive> BuildStrategy<Float, 1>::operator () (Iterable<Primitive> primitives, BoxConverter converter) {
+
+  using namespace Strategy1;
+
+  // The threshold hold at which a leaf is made in the BVH.
+  static constexpr uint32_t leaf_size = 4;
+
+  BuildStack todo;
+
+  const uint32_t Untouched    = 0xffffffff;
+  const uint32_t TouchedTwice = 0xfffffffd;
```

Comment:
I'm going to create an issue to better document how this works.

Reply:
Thanks! I was wondering about that